storage/client_range_test.go

```diff
@@ -150,3 +155,182 @@ func TestRejectFutureCommand(t *testing.T) {
 		t.Errorf("expected 15, got %v", v)
 	}
 }
+
+// TestTxnPutOutOfOrder tests a case where a put operation of an older
+// timestamp comes after a put operation of a newer timestamp in a
+// txn. The test makes sure such an out-of-order put succeeds and
+// overrides an old value.
+//
+// The test uses a "Writer" and a "Reader" to reproduce an
+// out-of-order put. The Writer executes a put operation in a txn and
+// commits it. The Reader executes two get operations on the same key
+// with higher priority (outside of a txn), which pushes the Writer's
+// intent forward in time, so that the Writer's retry happens in the
+// previous attempt's past.
+//
+// 1) The Writer executes a put operation with time T in a txn.
+// 2) Before the Writer's txn is committed, the Reader sends a get
+//    operation with time T+100. This pushes the Writer txn timestamp to
+//    T+100 and triggers the restart of the Writer's txn. A write intent
+//    at timestamp T+100 is written.
+// 3) Before the Writer starts a new epoch of the txn, the Reader sends
+//    another get operation with time T+200. The Writer txn timestamp is
+//    pushed again to T+200 by updating the timestamp of the write
+//    intent to T+200. We fail the actual get operation, so that
+//    the timestamp cache does not get updated.
+// 4) The Writer restarts its txn and executes the put operation
+//    again. This put operation becomes out-of-order since its
+//    timestamp is T+100 while the intent timestamp pushed at Step 3
+//    is T+200.
+// 5) The put operation still succeeds and overrides the old value.
+func TestTxnPutOutOfOrder(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	key := "key"
+
+	// Set up a filter to so that the get operation at Step 3 will return an error.
+	var numGets int32
+	numGets = 0
+	storage.TestingCommandFilter = func(args proto.Request, reply proto.Response) bool {
+		if _, ok := args.(*proto.GetRequest); ok &&
+			args.Header().Key.Equal(proto.Key(key)) &&
+			args.Header().Txn == nil {
+			// The Reader executes two get operations, each of which triggers two get requests
+			// (the first request fails and triggers txn push, and then the second request
+			// succeeds). Returns an error for the fourth get request to avoid timestamp cache
+			// update after the third get operation pushes the txn timestamp.
+			if atomic.AddInt32(&numGets, 1) == 4 {
+				reply.Header().SetGoError(util.Errorf("Test"))
+				return true
+			}
+		}
+		return false
+	}
+	defer func() {
```

Comment:
heh sorry i didn't mean move leaktest under it - take a look at other tests, these defers are usually set up early. leaktest should definitely be immediately after the method prototype